### PR TITLE
SSU: better fix for #658

### PR DESCRIPTION
**Pull request by MoroccanMalinois**
---
**By submitting this pull-request, I confirm the following:**

- I have read and understood the [Contributing Guide](https://github.com/anonimal/kovri-docs/blob/master/developer/CONTRIBUTING.md).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.
---
IMHO, i think this is a better fix for #658 "design-wise" (with the disclamer that i'm usually not very good at design)

IIUC, the idea behind the builder is that all data are already available and we only need to find pointers to these data. The problem is that the internal data of boost::asio::ip::address is not accessible, only a copy through `to_bytes().data()`. So i think, to be consistent design-wise, it is better to treat "ip" like other elements in the builder and leave the memory allocation outside. Or said otherwise, it was a bug in the usage of the class, not in the class itself ... i think ;) 

Ref #140

### CODE REVIEW

File: src/core/util/byte_stream.cc [hunk 1]
```diff
@@ -158,5 +158,20 @@ const std::string GetFormattedHex(const std::uint8_t* data, std::size_t size) {
...
+  std::unique_ptr<std::uint8_t[]> buffer =
+      std::make_unique<std::uint8_t[]>(*size);
+  std::memcpy(
+      buffer.get(),
+      address.is_v4() ? address.to_v4().to_bytes().data()
+                      : address.to_v6().to_bytes().data(),
+      *size);
+  return buffer;
+}
+
```
Comment (anonimal) [inline src/core/util/byte_stream.cc]: >std::unique_ptr<<std::uint8_t[]>>
>std::size_t* size)

At this point, why not refactor/implement/return a vector? Simply pass the address object, do the checks, create vector from `to_bytes().data()` (or `assign` if passing arg reference to *caller* vector object), and return the vector (*if* creating the object locally in the *callee*)? From there, `Insert` the vector data pointer and vector size. We could probably implement all of that in just a few lines (including the return) while reducing the session code by a couple of lines.

**Comment by MoroccanMalinois** [comment]
Thanks @anonimal. Ready

**Comment by MoroccanMalinois** [comment]
Just force pushed a mini update (start fixture variable name with m_)

File: src/core/util/byte_stream.cc [hunk 2]
```diff
@@ -158,5 +158,19 @@ const std::string GetFormattedHex(const std::uint8_t* data, std::size_t size) {
   return hex.str();
 }
 
+std::unique_ptr<std::vector<std::uint8_t>> AddressToByteVector(
```
Comment (anonimal) [inline src/core/util/byte_stream.cc]: I like this revision even though I had other thoughts in mind. Assuming we trust RVO, we may have also considered:

```cpp
std::vector<std::uint8_t> AddressToByteVector(
    const boost::asio::ip::address& address) 
{                                                   
  bool is_v4(address.is_v4());                      
  std::vector<std::uint8_t> vec(is_v4 ? 4 : 16);
  std::copy_n(
      vec.begin(),
      (is_v4 ? 4 : 16),
      (is_v4 ? address.to_v4().to_bytes().begin()
             : address.to_v6().to_bytes().begin()));
  return vec;                                       
}
```

or avoid the return:

```cpp
void AddressToByteVector(
    std::vector<std::uint8_t>& vec,
    const boost::asio::ip::address& address) 
{ 
  bool is_v4(address.is_v4());
  vec.resize((is_v4 ? 4 : 16));
  std::copy_n(
      vec.begin(), 
      (is_v4 ? 4 : 16),
      (is_v4 ? address.to_v4().to_bytes().begin()
             : address.to_v6().to_bytes().begin()));
}
```

but either of these would've required interface refactoring elsewhere (and I also prefer the move semantics of `unique_ptr`).

With that said, I think we should use our collaborated revision (https://github.com/monero-project/kovri/pull/671/commits/02ccfa87a1c996762ff752faba4794a5c3cf00b9) 👍

**Comment by MoroccanMalinois** [comment]
Thanks @anonimal. Ready

**Pull request merged**